src/compiler/parser.ts

```diff
@@ -5801,18 +5807,21 @@ namespace ts {
             return finishNode(node);
         }
 
-        function isIdentifierOrPattern() {
-            return token() === SyntaxKind.OpenBraceToken || token() === SyntaxKind.OpenBracketToken || isIdentifier();
+        function isIdentifierOrPrivateIdentifierOrPattern() {
+            return token() === SyntaxKind.OpenBraceToken
+                || token() === SyntaxKind.OpenBracketToken
+                || token() === SyntaxKind.PrivateIdentifier
+                || isIdentifier();
         }
 
-        function parseIdentifierOrPattern(): Identifier | BindingPattern {
+        function parseIdentifierOrPattern(privateIdentifierDiagnosticMessage?: DiagnosticMessage): Identifier | BindingPattern {

```

Comment:
I'm not a huge fan of optional parameters in the parser. Is there a lot of overhead to making this parameter required?

Reply:
Nope. Also forgot to pass them through to binding pattern parsing. Fixed.

Reply:
How do you know that they’re slow, by the way? I can’t see an appreciable difference running silly contrived test cases in Chrome.

Reply:
Discussions around the office, which is why I'm not convinced it's true. I may have misunderstood something, or it may be out of date, or the impact may never be that big.

Reply:
I'm not certain whether this has changed in the past few years, but the original rationale was this:

When V8 compiles a function, that compiled function has a fixed number of arguments. If you undersupply arguments to that function, what V8 does is creates a *new* compiled function with that number of arguments that in turn calls the original compiled function, with `undefined` filling in the missing arguments. The reason this can degrade performance is due to the fact that V8 has to generate this new function, and that new function could possibly be garbage collected at some point if it is used infrequently and may eventually need to be generated again.

I *have* seen this issue in practice several years ago, as there was a bug andy-ms found regarding performance in one of our core functions, though I cannot find the bug report or related PR at this time.

Reply:
@rbuckton do you know if/how default parameters affect that? (I guess it’s a moot point for us since we downlevel to ES5.)

Reply:
Not offhand, no. @syg might know more about whether this optimization still exists (or ever existed?) or could at least point us to someone who might know more.